File: 567-permutation-in-string/permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int l = 0;
        int r = 0;
        int window_size = s1.size();
        std::unordered_map<char,int> freq_mp;
        std::unordered_map<char,int> window_freq_mp;
        for (int i = 0; i < s1.size(); i++) freq_mp[s1[i]]++;

        while(r<s2.size()){
            window_freq_mp[s2[r]]++;
            if (r >= window_size) {
                char left_char = s2[l];
                window_freq_mp[left_char]--;
                if (window_freq_mp[left_char] == 0) {
                    window_freq_mp.erase(left_char);
                }
                l++;
            }
            if(window_freq_mp==freq_mp) return true;
            r++;
        }
        return false;
        
    }
};

File: 567-permutation-in-string/permutation-in-string.py
class Solution:
    # O(n*mlogm)solution
    # def checkInclusion(self, s1: str, s2: str) -> bool:
    #     s1_sorted = sorted(s1)
    #     for i in range(len(s2)-len(s1)+1):
    #         if sorted(s2[i:i+len(s1)])== s1_sorted:
    #             return True
    #     return False

    def checkInclusion(self, s1: str, s2: str) -> bool:
        s1_freq = Counter(s1)
        left, right = 0, 0
        window_freq = {}
        
        while(right<len(s2)):
            window_freq[s2[right]] = window_freq.get(s2[right],0)+1
            window_size = right-left+1            
            if window_size > len(s1):
                window_freq[s2[left]]-=1
                if window_freq[s2[left]]==0: window_freq.pop(s2[left])
                left+=1
            if window_freq==s1_freq: return True
            right+=1
        
        return False

                

        